### PR TITLE
feat: ConnectionOptions<T> can convert to `internal::Options`

### PR DESCRIPTION
**Pull request by devjgm**
Part of https://github.com/googleapis/google-cloud-cpp/issues/5738

This give us and users a path to migrate from the old `ConnectionOptions<T>` instances so the new `google::cloud::Options` class. After this PR, we can start adding overloads to existing functions that take `ConnectionOptions<T>` to also take `Options`, and have the `Options` overload be preferred.

## Attention

Since this is the first PR where we're really _using_ the `Options` class, I'll make a few observations and ask for feedback.

1. I don't love that `opts.set<GrpcUserAgentPrefixOption>(std::set<std::string>{user_agent_prefix_});` requires the user to spell the `std::set` part. The reason for this is because `Options::set<T>()` takes a variadic forwarding reference, and passing _only_ an initializer list (e.g., `opts.set<GrpcUserAgentPrefixOption>({user_agent_prefix_});`) will not be able to deduce `U`. We _could_ "fix" this if `Options` takes a position that all option structs _must_ have one data member named `.value` (which we're currently agnostic about). In this case, we could use the type of the `.value` member to _be_ the type of set's argument, thus avoiding the forwarding reference, and thus enabling caller-side initializer lists. Thoughts?
2. `Options` only "gets" _values_, so that requires a copy. This is fine and easy, but it could become an issue if these options become large and get copied a lot. For example, consider an option whose value is a set/map, and the caller wants to insert a value. That would require making a copy, inserting the value, then moving the new value over the old one. Just something to consider.
3. The endpoint option is gRPC specific: `GrpcEndpointOption`. Is that fine? I think it's OK. If we make it transport agnostic, then we'll need to find another header for it to live in.
4. `GrpcUserAgentPrefixOption` has a _std::set_ for all the prefixes. This means that the user can't control the order of the prefixes. Is this OK with us? I think it's probably fine. I don't know if/why the order would matter. If we do care about the order, we could use a `std::deque`.

Note: I did not change the implementation of `ConnectionOptions<T>` to simply hold a single `google::cloud::Options` instance and re-implement its getters/setters in terms of `Options`, because there were enough little semantic differences that made it messy. For example, `channel_pool_domain()` is now implemented as a single key in `GrpcChannelArgumentsOption`, which makes it very difficult for the `channel_pool_domain()` getter to return a const-reference (to what)? I could stuff a default empty string in there for that option, but then an empty channel_pool_domain is used as a signal to not set that channel attribute.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864)

### CODE REVIEW

**Review by coryan — COMMENTED** [review]
Need to leave, some comments below, but I did not complete the review.

> Reviewed 7 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-:-MTl20t-2DDIXsbNPMVt:b-i2fdhw)* status: 7 of 8 files reviewed, 4 unresolved discussions (waiting on @devjgm)

___
*[google/cloud/connection_options.h, line 60 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1-776RxS90eIRwNF:-MTl1-776RxS90eIRwNG:bcrl1bf) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/connection_options.h#L60)):*
> ```objc
> 
>   /// Moves all the settings in this class into an `google::cloud::Options`.
>   Options ToOptions() && {
> ```

can we make this a friend in `internal::`?  We should try to not expose APIs that we need for our implementation in the public APIs of other classes.
___
*[google/cloud/connection_options.h, line 74 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl18X4-cGETLlHBEWI:-MTl18X4-cGETLlHBEWJ:b-roh631) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/connection_options.h#L74)):*
> ```objc
>     if (!channel_pool_domain_.empty()) {
>       opts.set<GrpcChannelArgumentsOption>(std::map<std::string, std::string>{
>           {GRPC_ARG_CHANNEL_POOL_DOMAIN, std::move(channel_pool_domain_)}});
> ```

This macro does not exist in sufficiently old versions of gRPC, see the Debian buster build ...
___
*[google/cloud/grpc_options.h, line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1hlNBO-GB3Kkvdlb:-MTl1hlNBO-GB3Kkvdlc:bqzadwo) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/grpc_options.h#L67)):*
> ```objc
>  * number of users running particular versions of their system or library.
>  */
> struct GrpcUserAgentPrefixOption {
> ```

User agent prefixes apply to both gRPC and REST clients.
___
*[google/cloud/grpc_options.h, line 96 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1o4TBoSb5Lttcijp:-MTl1o4TBoSb5Lttcijq:bttltn1) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/grpc_options.h#L96)):*
> ```objc
>  * - rpc-streams
>  */
> struct GrpcTracingComponentsOption {
> ```

I think this can be generalized too, feel free to push these generalizations to a future PR, but be aware of the API-breakage that might cause.

**Review by coryan — COMMENTED** [review]
Sorry for the incomplete review, I had to run to do an interview. Some suggestions below.  It might be easier to start with these options in the `internal::` namespace` to avoid breaking APIs as we stabilize the design for all libraries.

If you agree with these changes, but would prefer to do them in future PRs that works too.

> Reviewed 1 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-:-MTlHHZ10wNR9_5AHCGo:buug4id)* status: all files reviewed, 4 unresolved discussions (waiting on @devjgm)

___
*[google/cloud/grpc_options.h, line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1hlNBO-GB3Kkvdlb:-MTlH3zhAH5GO8aF01Wg:b-i36k6o) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/grpc_options.h#L67)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

User agent prefixes apply to both gRPC and REST clients.
</blockquote></details>

Consider refactoring to a common place.

**Review by devjgm — COMMENTED** [review]
Done. Moved all the `Options`-related code into `internal::`.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-:-MTqciP_-k5zAvBjpzV8:b60oy2y)* status: 0 of 12 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[google/cloud/connection_options.h, line 60 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1-776RxS90eIRwNF:-MTqcpP88gbidLkuDp_i:bawgmoz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/connection_options.h#L60)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

can we make this a friend in `internal::`?  We should try to not expose APIs that we need for our implementation in the public APIs of other classes.
</blockquote></details>

Moved to a non-member in `internal::`. Note: That once we have finalized the `Options` API and class, I think we should consider moving this conversion function back into the class. I think a user-visible `ToOptions()` function would be useful for users so that they can easily migrate from the old `ConnectionOptions` -> the new `Options`.
___
*[google/cloud/connection_options.h, line 74 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl18X4-cGETLlHBEWI:-MTqd7uS6WlRN5ILyJpW:b-ab1zf2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/connection_options.h#L74)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This macro does not exist in sufficiently old versions of gRPC, see the Debian buster build ...
</blockquote></details>

Thanks. Moved back to using the string and I kept the original comment that was there. I thought our gRPC requirements had moved forward enough that this macro would be available everywhere, so I wanted to check that. I guess not :-)
___
*[google/cloud/grpc_options.h, line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1hlNBO-GB3Kkvdlb:-MTqdLiY-gsyCNvECB5H:bvvrisb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/grpc_options.h#L67)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider refactoring to a common place.
</blockquote></details>

Moved to a `common_options.h` header.
___
*[google/cloud/grpc_options.h, line 96 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-MTl1o4TBoSb5Lttcijp:-MTqdUll2nxysicuojcu:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/452c0b3bdde038a793175368a433e0853cbcffc9/google/cloud/grpc_options.h#L96)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think this can be generalized too, feel free to push these generalizations to a future PR, but be aware of the API-breakage that might cause.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 12 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5864#-:-MTqgVnFCm7M-xAIdra0:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

File: google/cloud/internal/common_options.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#include <string>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+
+/**
+ * Change the endpoint.
+ *
+ * In almost all cases the a suitable default will be chosen automatically.
```
Comment (devbww) [inline google/cloud/internal/common_options.h:30]: s/ the//

File: google/cloud/internal/common_options.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * Applications may need to be changed to (1) test against a fake or simulator,
+ * or (2) to use a beta or EAP version of the service.
```
Comment (devbww) [inline google/cloud/internal/common_options.h:32]: "changed to (1) test ..., or (2) use a ..."

OR

"changed (1) to test ..., or (2) to use a ..."

File: google/cloud/internal/grpc_options.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_GRPC_OPTIONS_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_GRPC_OPTIONS_H
+
+#include "google/cloud/background_threads.h"
+#include "google/cloud/internal/options.h"
+#include "google/cloud/tracing_options.h"
+#include "google/cloud/version.h"
+#include <grpcpp/grpcpp.h>
+#include <map>
+#include <set>
```
Comment (devbww) [inline google/cloud/internal/grpc_options.h:24]: Unused?

**Comment by devjgm** [comment]
Addressed in follow-up PR https://github.com/googleapis/google-cloud-cpp/pull/5875

**Pull request merged**